src/execution_plan/execution_plan_reduce_to_apply.c

```diff
@@ -0,0 +1,164 @@
+#include "execution_plan.h"
+#include "ops/ops.h"
+#include "../query_ctx.h"
+#include "../ast/ast_mock.h"
+#include "./optimizations/optimizer.h"
+
+static OpBase *_buildMatchBranch(ExecutionPlan *plan, const cypher_astnode_t *path) {
+	AST *ast = QueryCtx_GetAST();
+
+	rax *bound_vars = NULL;
+	const char **arguments = NULL;
+	if(plan->root) {
+		bound_vars = raxNew();
+		// Rather than cloning the record map, collect the bound variables along with their
+		// parser-generated constant strings.
+		ExecutionPlan_BoundVariables(plan->root, bound_vars);
+		// Prepare the variables for populating the Argument ops we will build.
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+
+	// Initialize an ExecutionPlan that shares this plan's Record mapping.
+	ExecutionPlan *match_branch_plan = ExecutionPlan_NewEmptyExecutionPlan();
+	match_branch_plan->record_map = plan->record_map;
+
+	// If we have bound variables, build an Argument op that represents them.
+	if(arguments) match_branch_plan->root = NewArgumentOp(plan, arguments);
+
+	// Build a temporary AST holding a MATCH clause.
+	AST *match_branch_ast = AST_MockMatchPattern(ast, path);
+
+	ExecutionPlan_PopulateExecutionPlan(match_branch_plan, NULL);
+
+	AST_MockFree(match_branch_ast);
+	QueryCtx_SetAST(ast); // Reset the AST.
+
+	OpBase *branch_match_root = match_branch_plan->root;
+	ExecutionPlan_BindPlanToOps(branch_match_root, plan);
+	OpBase *argument = ExecutionPlan_LocateFirstOp(branch_match_root, OPType_ARGUMENT);
+	if(argument) {
+		if(argument->modifies) {
+			/* Don't lose information when optimizing this branch. The optimizations are lookgin for
+			 * specific operation types and if the result of the operations is achievable by one of its
+			 * decedents, this op is redundant. The argument op "modifies" the execution plan by
+			 * showing the bounded arguments. Given that, once optimized the label scan op is
+			 * so called redundant and will be removed, since its result is achievable by the argument op.
+			 * But this is wrong since the argument op is not modifing anything.
+			 * The next sequence should avoid the removal of the label scan op, and yet optimize the rest
+			 * of the branch. */
+			if(argument->parent->type == OPType_NODE_BY_LABEL_SCAN) {
+				OpBase *parent = argument->parent;
+				// Temporary remove label scan op. Now argument op is the top op.
+				ExecutionPlan_RemoveOp(plan, parent);
+				// Optimize the rest of the plan. Unfortunately, this currently needs to be done in place.
+				optimizePlan(match_branch_plan);
+				// Add back the scan op in its proper place.
+				ExecutionPlan_ReplaceOp(match_branch_plan, argument, parent);
+				ExecutionPlan_AddOp(parent, argument);
```

Comment:
Is all of this logic necessary? Can't you just perform the steps of freeing and NULL-setting the `argument->modifies` array so that it doesn't trigger invalid optimizations? I don't think the calls to remove/replace ops are required.

Reply:
no
for queries like
`MATCH (a), (b) WHERE a.id = 0 AND (a)-[:T]->(b:TheLabel) OR (a)-[:T*]->(b:MissingLabel) RETURN DISTINCT b`
On one hand, you want to remove the label scan operation that will be generated for the branch `(a)-[:T*]->(b:MissingLabel)`, on the other hand, you want a variable-length traversal with an expand-into feature. We want the optimization but the argument modifies causes bad execution for that.